PR benchmarks

sebastienros opened the pull request:
Submitting for comments. I am thinking of creating a separate file that will contain which "targets" we want to support, which will include the components to builds, and the scenarios to execute. Right now it supports `/benchmark kestrel` and `/benchmark mvc`.

I also think a comment should be created to acknowledge the benchmark is running with a link to the running workflow like the backport command is doing. 

The results look like this right now:

Arguments:
```
--config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/plaintext.benchmarks.yml --scenario plaintext --application.framework net6.0 --application.options.outputFiles ./artifacts/bin/Microsoft.AspNetCore.Server.Kestrel/release/net6.0/
```
Results:
```

| application           | base                  | head                  |         |
| --------------------- | --------------------- | --------------------- | ------- |
| CPU Usage (%)         |                    99 |                    99 |   0.00% |
| Cores usage (%)       |                   397 |                   396 |  -0.25% |
| Working Set (MB)      |                    91 |                    93 |  +2.20% |
| Private Memory (MB)   |                   303 |                   311 |  +2.64% |
| Build Time (ms)       |                23,507 |                 4,342 | -81.53% |
| Start Time (ms)       |                   293 |                   295 |  +0.68% |
| Published Size (KB)   |               108,558 |               108,558 |   0.00% |
| .NET Core SDK Version | 6.0.100-rc.1.21417.19 | 6.0.100-rc.1.21417.19 |         |


| load                   | base       | head       |         |
| ---------------------- | ---------- | ---------- | ------- |
| CPU Usage (%)          |         57 |         59 |  +3.51% |
| Cores usage (%)        |        227 |        237 |  +4.41% |
| Working Set (MB)       |         40 |         42 |  +5.00% |
| Private Memory (MB)    |        358 |        358 |   0.00% |
| Start Time (ms)        |          0 |          0 |         |
| First Request (ms)     |        167 |        225 | +34.73% |
| Requests/sec           |    783,294 |    816,098 |  +4.19% |
| Requests               | 11,824,795 | 12,321,760 |  +4.20% |
| Mean latency (ms)      |       3.14 |       3.14 |   0.00% |
| Max latency (ms)       |      50.25 |      82.87 | +64.92% |
| Bad responses          |          0 |          0 |         |
| Socket errors          |          0 |          0 |         |
| Read throughput (MB/s) |      98.60 |     102.73 |  +4.19% |
| Latency 50th (ms)      |       2.84 |       2.73 |  -3.87% |
| Latency 75th (ms)      |       4.21 |       4.05 |  -3.80% |
| Latency 90th (ms)      |       5.44 |       5.28 |  -2.94% |
| Latency 99th (ms)      |      11.29 |      12.63 | +11.87% |
```

sebastienros committed:
Create PR benchmarks workflow

sebastienros commented:
How it works:

1. A comment starting with `/benchmark ` is detected, and the next part is exctracted (the target)
2. The user who created the comment is checked against the repo collaborators to prevent external users from triggering builds on malicious PRs
3. The `crank` dotnet tool is installed
4. The PR base branch is checked out, and the specified target defines which component to build, e.g. Kestrel, MVC
5. A crank benchmark is executed with these binaries and stored (base.json)
6. The PR code is checked out, merged to the target branch, and built
7. A crank benchmark is executed with these binaries and stored (head.json)
8. The `crank compare` command is invoked on the results, and added to a comment on the initial PR

No artifacts are stored, so built assets can't be served publicly.
The current target machines are on Azure and isolated in a private network.
The communication to the Azure machine is done through an Azure Relay Hybrid Connection. Its connection string needs to be stored in a GitHub's repository secret, which can be revoked or rotated.

BrennanConroy commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,189 @@
+          const repo_name = context.payload.repository.name;
+          const comment_user = context.payload.comment.user.login;
+
+          var membership = await github.repos.checkCollaborator({
```
Shouldn't this be verifying the return value?

sebastienros replied:
It's throwing when it's wrong

BrennanConroy commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,189 @@
+    # comments can be triggered on issues and prs, ensure this is a PR
+    if: github.event.issue.pull_request != '' && contains(github.event.comment.body, '/benchmark')
+    # Comment out this line to test the scripts in a fork
+    # if: github.repository == 'dotnet/aspnetcore'
```
fix

BrennanConroy commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,189 @@
+          echo "Building Mvc"
+          cd ./src/Mvc
+          ./build.sh -c release -NoBuildNodeJS
+        fi
```
else error

sebastienros committed:
Update benchmark.yml

sebastienros commented:
Updated. I split each command in it's own step, which will allow to create custom sets of benchmarks generating custom outputs. For instance with profile matrices, or multiple scenarios, and be able to generate custom tables for instance. Since each benchmark stores the results in json, we can then use `jq` to extract which metrics to read in each run and build custom tables. The current example only show how to use `crank compare`.

sebastienros committed:
Render error when user is not a collaborator

BrennanConroy commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,191 @@
+          const pull_request = (await github.request(context.payload.issue.pull_request.url)).data;
+          const target_branch = pull_request.base.ref;
+          const source_repository = pull_request.head.repo.clone_url;
+          const source_branch = pull_request.head.ref;
```
These are all unused

BrennanConroy commented on .github/workflows/benchmark.yml:
````diff
@@ -0,0 +1,191 @@
+        crank $CRANK_CONFIGS --profile $CRANK_PROFILE --relay AZURE_RELAY --json "$GITHUB_WORKSPACE/head.json" $CRANK_ARGUMENTS
+
+        cd $GITHUB_WORKSPACE
+        echo "Arguments:"$'\n```\n'$"$CRANK_ARGUMENTS"$'\n```\n' >> results.txt
````
Should this bit be put in a separate step so it's used by all benchmarks?

sebastienros replied:
This is intentional as I expect some commands to diverge from this and call crank multiple times, or generate an output that is different. So just repeating to simplify authoring new commands.

sebastienros committed:
Remove unused variables

sebastienros committed:
Add yarp command

sebastienros commented:
@wtgodbe I could use a quick look from you in case you see things that are not permitted, or other concerns with github actions.

wtgodbe commented on .github/workflows/benchmark.yml:
```diff
@@ -0,0 +1,220 @@
+
+          core.info(`Verified ${comment_user} is a repo collaborator.`);
+
+          // Verify this action is and PR comment
```
"is and" -> "is a"

sebastienros replied:
```suggestion
          // Verify this action is a PR comment
```

wtgodbe reviewed (changes requested):
This looks good to me in terms of what it's doing, but it'd be good to add a doc listing how to trigger it

sebastienros committed:
Update .github/workflows/benchmark.yml

sebastienros committed:
Add documentation

sebastienros committed:
Typos

captainsafia commented on docs/PRBenchmarks.md:
```diff
@@ -0,0 +1,22 @@
+# Pull Request Benchmarks
+
+Pull requests that may impact performance should be benchmarked to ensure no regression is introduced.
+The most common way is to build the PR on your development machine and submit the binaries to an existing [scenario](https://github.com/aspnet/Benchmarks/tree/main/scenarios). To simplify this process this repository can do it automatically using GitHub Actions.
```
```suggestion
The most common way is to build the PR on your development machine and submit the binaries to an existing [scenario](https://github.com/aspnet/Benchmarks/tree/main/scenarios). For simplicity, this process can be done automatically via GitHub Actions.

```

captainsafia commented on docs/PRBenchmarks.md:
```diff
@@ -0,0 +1,22 @@
+
+## Submitting a PR for benchmarking
+
+A special set of comments are recognized by this repository in order to trigger a benchmark. The comment starts with `/benchmark` and takes the name of the benchmark command as an argument. For instance, `/benchmark kestrel` will build the Kestrel components and submit a `plaintext` scenario benchmark to dedicated machines.
```
Where does one find the list of benchmarks?

sebastienros replied:
Two lines bellow ;)

sebastienros committed:
Update docs/PRBenchmarks.md

Co-authored-by: Safia Abdalla <safia@microsoft.com>
